src/main/java/pl/put/poznan/transformer/logic/text/numbers/NumberExpandText.java

```diff
@@ -0,0 +1,150 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Expand numbers to text
+ */
+public class NumberExpandText extends TextTransformer {
+    private static final Logger logger = LoggerFactory.getLogger(NumberExpandText.class);
+
+    public NumberExpandText(Text text){
+        super(text);
+    }
+
+    public static String name = "expand";
+
+    @Override
+    public String transform() {
+        return setNumberExpand(text.transform());
+    }
+
+    private String setNumberExpand(String text) {
+        ArrayList<String> matches = MatchPattern(text);
+
+        try {
+            ArrayList<String> converted = ConvertNumberToText(matches);
+            return ReplaceNumbers(text, converted);
+        } catch (IOException | ParseException e) {
+            logger.warn("Can't expand number:", e);
+            return text;
+        }
+    }
+
+    /**
+     * Matches pattern in given text
+     * @return all pattern matches in list
+     */
+    public static ArrayList<String> MatchPattern(String text) {
+        String regex = "[+-]?([0-9]+\\.?[0-9]*|\\.[0-9]+)";
+        Pattern pattern = Pattern.compile(regex);
+
+        Matcher matcher = pattern.matcher(text);
+        ArrayList<String> matches = new ArrayList<>();
+        while(matcher.find()) {
+            matches.add(matcher.group());
+        }
+        return matches;
+    }
+
+    /**
+     * Converts numbers to text
+     * @param matches is a given list with numbers as strings
+     * @throws IOException
+     * @throws ParseException
+     * @return list of numbers converted to text
+     */
+    public static ArrayList<String> ConvertNumberToText(ArrayList<String> matches) throws IOException, ParseException {
+        ArrayList<String> converted = new ArrayList<>();
+        Object object = new JSONParser().parse(new FileReader("src/main/resources/NumberToText.json"));
+        JSONObject jo = (JSONObject) object;
+
+        int lenbeforedot, lenafterdot;
+        for (String match : matches) {
+            lenbeforedot = lenafterdot = 0;
+            if (match.endsWith(".")) {
+                match = match.substring(0, match.length() - 1);
+                lenbeforedot = match.length();
+            } else if (match.contains(".")) {
+                String[] parts = match.split("\\.");
+                lenbeforedot = parts[0].length();
+                lenafterdot = parts[1].length();
+            } else {
+                lenbeforedot = match.length();
+            }
+
+            int number;
+            ArrayList<String> numbers;
+            StringBuilder result = new StringBuilder();
+            for (int i = lenbeforedot, place = 0; i > 0; --i, ++place) {
+                number = match.charAt(place) - '0';
+                if (i == 1) {                                           // 0-9
+                    if (number != 0) {                                      // 1-9
+                        numbers = (ArrayList<String>) jo.get("ones");
```

Comment:
Kłują mnie te operacje na jsonie i castowanie. Osobiście widziałbym to raczej w formie zmapowania jsona na obiekt jakiejś naszej klasy, z odpowiednimi polami. Coś takiego zapewnia np [fasterxml](https://www.baeldung.com/jackson-object-mapper-tutorial) albo [jsonb](https://www.samouczekprogramisty.pl/format-json-w-jezyku-java/).